File: decision_room/web/static/dossier.js
/* Consultable owner memory; every mutation uses the same backend as conversations. */
const factKinds = {context: "Información", priority: "Prioridad", definition: "Definición", availability: "Disponibilidad", open_question: "Duda", result_reference: "Referencia a informe"};
const factStates = {declared: "Declarado por ti", proposed: "Pendiente de confirmar", conflicted: "Contradicción pendiente", withdrawn: "Retirado", superseded: "Versión anterior"};
function factScope(content, dossier) {
  if (content.scope === "business") return "Todo el negocio";
  const dataset = dossier.datasets.find(d => d.id === content.scope_id || (d.files || []).some(f => f.id === content.scope_id));
  const file = dataset?.files?.find(f => f.id === content.scope_id);
  return file ? file.name : dataset ? dataset.title : "Fuente o conjunto específico";
}
function validity(content) {
  return content.temporal_scope === "unresolved" ? "Fecha por aclarar" : content.valid_from || content.valid_until ? `${content.valid_from || "Sin inicio indicado"} → ${content.valid_until || "Sin fin indicado"}` : "Sin periodo indicado";
}
function originDetails(fact) {
  return `<details class="fact-origin"><summary>Origen y alcance</summary><p>${esc(fact.origin_kind === "profile" ? "Perfil del negocio" : fact.origin_key?.startsWith("chat_message:") ? "Conversación" : fact.origin_kind === "manual" ? "Declaración o edición explícita" : "Aclaración de un análisis")} · ${fmtDate(fact.created_at)}</p><blockquote>${esc(fact.original_text || fact.quote)}</blockquote>${fact.question ? `<p>Pregunta: ${esc(fact.question)}</p>` : ""}${fact.conversation_id ? `<a href="#chat/${esc(fact.conversation_id)}">Abrir conversación</a>` : ""}</details>`;
}
function dossierGroupFor(fact) {
  if (fact.status !== "declared" || fact.content.kind === "open_question") return "review";
  if (fact.content.kind === "priority") return "preferences";
  if (fact.content.kind === "definition" || fact.content.scope !== "business") return "data";
  const topic = fact.content.topic || "";
  if (fact.content.kind === "availability" && /open|hour|schedule|horario|apertura|cierre/.test(topic)) return "business";
  if (fact.content.kind === "availability" || /csv|data|dataset|file|source|column|row|field|revenue|sales|cost|ticket|venta|dato|archivo|unidad|importe/.test(topic)) return "data";
  return "business";
}
function dossierFactRow(f, dossier) {
  const unresolved = f.status !== "declared";
  return `<article class="dossier-fact-row"><div class="dossier-fact-main"><p class="fact-statement">${esc(f.content.statement)}</p><button class="dossier-quick-edit" type="button" data-edit="${esc(f.fact_id)}">${f.status === "conflicted" ? "Resolver" : "Editar"}</button></div>${unresolved ? `<p class="dossier-fact-state">${esc(factStates[f.status])}</p>` : ""}<details class="dossier-fact-more"><summary aria-label="Ver detalles y origen">···</summary><p>${esc(factKinds[f.content.kind])} · ${esc(factScope(f.content, dossier))} · ${esc(validity(f.content))}</p>${f.alternatives?.length ? `<details><summary>Versiones en conflicto</summary>${f.alternatives.map(a => `<p>${esc(a.content?.statement || a.statement || a.quote || "Consulta la declaración original.")}</p>`).join("")}</details>` : ""}${originDetails(f)}<div class="dossier-actions">${f.status === "proposed" && f.content.temporal_scope !== "unresolved" ? `<button class="button secondary" data-confirm="${esc(f.fact_id)}">Confirmar</button>` : ""}<button class="text-button" data-withdraw="${esc(f.fact_id)}">Retirar</button></div></details><p class="field-help" role="status" id="fact-status-${esc(f.fact_id)}"></p></article>`;
}
function dossierInformationMarkup(dossier, facts) {
  const b = dossier.business;
  const groups = [
    ["review", "Por revisar", "Datos que necesitan una decisión"],
    ["business", "Sobre el negocio", "Identidad, actividad y funcionamiento"],
    ["preferences", "Preferencias y objetivos", "Lo que quieres conseguir o recordar"],
    ["data", "Datos y definiciones", "Archivos, disponibilidad e interpretación"],
  ];
  const grouped = Object.fromEntries(groups.map(([key]) => [key, facts.filter(f => dossierGroupFor(f) === key)]));
  const sections = groups.filter(([key]) => grouped[key].length).map(([key, title, hint]) => {
    const open = key === "review" || key === "business";
    return `<details class="dossier-group" data-default-open="${open}" ${open ? "open" : ""}><summary><span><strong>${title}</strong><small>${hint}</small></span><span class="dossier-group-count">${grouped[key].length} ${grouped[key].length === 1 ? "dato" : "datos"}</span></summary><div class="dossier-fact-list">${grouped[key].map(f => dossierFactRow(f, dossier)).join("")}</div></details>`;
  }).join("");
  return `<section class="dossier-overview"><div class="dossier-overview-head"><div><p class="eyebrow">PRESENTACIÓN</p><h2>${esc(b.name)}</h2></div><a href="#business">Editar presentación</a></div><p class="business-description dossier-description-preview">${esc(b.description || "Añade contexto a tu ritmo; no hay campos obligatorios que completar.")}</p>${(b.description || "").length > 220 ? `<details class="dossier-description-full"><summary>Ver descripción completa</summary><p>${esc(b.description)}</p></details>` : ""}</section><div class="dossier-information-head"><div><h2>Información</h2><p>Encuentra y corrige lo que el agente sabe de tu negocio.</p></div><button class="button secondary" id="fact-new">Añadir información</button></div>${facts.length ? `<label class="dossier-search">Buscar información<input id="dossier-fact-search" type="search" placeholder="Buscar en tu negocio…"></label><div class="dossier-groups">${sections}</div><p class="dossier-empty" id="dossier-search-empty" hidden>No hay datos que coincidan con la búsqueda.</p>` : '<p class="dossier-empty">No hay información activa. Puedes añadirla o contarla en una conversación.</p>'}`;
}
async function dossierPage(tab = "information") {
  if (!state.business) { businessForm(true); return; }
  const generation = state.generation, business = state.business.id;
  const dossier = await api("/api/business/dossier");
  if (generation !== state.generation) return;
  if (dossier.business_id !== business) { await route(); return; }
  state.business = dossier.business;
  state.memory = dossier.memory;
  const render = (selected) => {
    const b = state.business;
    const facts = dossier.facts.filter(f => !["withdrawn", "superseded"].includes(f.status));
    shell(`<div class="page-heading"><div><p class="eyebrow">TU MEMORIA COMPARTIDA</p><h1>Mi negocio</h1><p>Revisa lo que recuerda el agente y los datos que puede consultar.</p></div><button class="button secondary" id="dossier-refresh">Actualizar ficha</button></div><div class="dossier-tabs" role="group" aria-label="Secciones de Mi negocio">${[["information", "Información"], ["data", "Datos y archivos"], ["history", "Cambios"]].map(([key, label]) => `<button class="button ${selected === key ? "primary" : "secondary"}" aria-pressed="${selected === key}" data-dossier-tab="${key}">${label}</button>`).join("")}</div><div id="dossier-content">${selected === "information" ? dossierInformationMarkup(dossier, facts) : selected === "data" ? datasetMarkup(dossier) : `<h2>Historial de cambios</h2><p>Se conservan las revisiones anteriores. Una información retirada no vuelve a utilizarse automáticamente.</p>${dossier.history.length ? dossier.history.map(f => `<article class="dossier-card"><div class="dossier-meta"><span>${fmtDate(f.created_at)} · Revisión ${f.revision}</span><span>${esc(factStates[f.status])}${dossier.facts.some(x => x.fact_id === f.fact_id && x.revision > f.revision) ? " · Histórica" : ""}</span></div><p>${esc(f.content.statement)}</p><p class="field-help">${esc(factScope(f.content, dossier))} · ${esc(validity(f.content))}${f.change_kind === "future" ? " · Cambio desde una fecha" : ""}</p>${originDetails(f)}${f.status === "withdrawn" && dossier.facts.some(x => x.fact_id === f.fact_id && x.revision === f.revision) ? `<button class="button secondary" data-edit="${esc(f.fact_id)}">Restaurar mediante corrección</button>` : ""}</article>`).join("") : '<p class="dossier-empty">Aún no hay cambios registrados.</p>'}`}</div><div id="dossier-editor"></div><p class="field-help"><a href="#businesses">Cambiar de negocio</a></p>`, "my-business", "Mi negocio");
    document.querySelectorAll("[data-dossier-tab]").forEach(button => button.onclick = () => render(button.dataset.dossierTab));
    document.querySelector("#dossier-refresh").onclick = () => refresh(selected);
    document.querySelector("#fact-new")?.addEventListener("click", () => editFact(null));
    document.querySelector("#dossier-fact-search")?.addEventListener("input", (event) => {
      const query = event.target.value.normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase().trim();
      let matches = 0;
      document.querySelectorAll(".dossier-group").forEach(group => {
        let visible = 0;
        group.querySelectorAll(".dossier-fact-row").forEach(row => {
          const text = row.querySelector(".fact-statement").textContent.normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase();
          row.hidden = !text.includes(query);
          if (!row.hidden) visible++;
        });
        group.hidden = visible === 0;
        group.open = query ? visible > 0 : group.dataset.defaultOpen === "true";
        matches += visible;
      });
      document.querySelector("#dossier-search-empty").hidden = matches !== 0;
    });
    document.querySelectorAll("[data-edit]").forEach(button => button.onclick = () => editFact(dossier.facts.find(f => f.fact_id === button.dataset.edit)));
    for (const action of ["confirm", "withdraw"]) document.querySelectorAll(`[data-${action}]`).forEach(button => {
      const fact = dossier.facts.find(f => f.fact_id === button.dataset[action]);
      const requestKey = crypto.randomUUID();
      button.onclick = async () => {
        button.disabled = true;
        const status = document.querySelector(`#fact-status-${fact.fact_id}`);
        try {
          await api("/api/business/memory", {method: "POST", body: {business_id: business, action, fact_id: fact.fact_id, expected_revision: fact.revision, request_key: requestKey}});
          if (generation !== state.generation) return;
          toast(action === "withdraw" ? "Información retirada de la memoria compartida." : "Información confirmada.");
          await refresh(selected);
        } catch (error) { if (status?.isConnected) status.textContent = error.message; button.disabled = false; }
      };
    });
    if (selected === "data") bindDatasetUpload(dossier, business, generation, () => refresh("data"));
  };
  const refresh = async (selected) => {
    try { await dossierPage(selected); } catch (error) { if (generation === state.generation) toast(error.message); }
  };
  const editFact = (fact) => {
    const c = fact?.content || {kind: "context", statement: "", scope: "business", scope_id: null, valid_from: null, valid_until: null, temporal_scope: "unspecified", result_id: null};
    const scopes = [["business", "Todo el negocio"], ...dossier.datasets.flatMap(d => [[`analysis:${d.id}`, `${d.title} · versión ${d.version}`], ...(d.files || []).map(f => [`source:${f.id}`, `Archivo: ${f.name}`])])];
    const target = c.scope === "business" ? "business" : `${c.scope}:${c.scope_id}`;
    const editor = document.querySelector("#dossier-editor");
    editor.innerHTML = `<section class="dossier-card"><h2>${fact ? "Corregir información" : "Añadir información"}</h2><form id="fact-form" class="dossier-form">${!fact ? '<label>Concepto<input name="concept" required maxlength="70" placeholder="Por ejemplo, margen objetivo"></label>' : ""}<label>Tipo<select name="kind">${Object.entries(factKinds).filter(([k]) => k !== "result_reference" || c.kind === k).map(([key,label]) => `<option value="${key}" ${key === c.kind ? "selected" : ""}>${label}</option>`).join("")}</select></label><label>Información<textarea name="statement" required maxlength="1600" rows="3">${esc(c.statement)}</textarea></label><label>Se aplica a<select name="scope">${scopes.map(([key,label]) => `<option value="${esc(key)}" ${target === key ? "selected" : ""}>${esc(label)}</option>`).join("")}</select></label><div class="dossier-dates"><label>Desde (opcional)<input type="date" name="valid_from" value="${esc(c.valid_from || "")}"></label><label>Hasta (opcional)<input type="date" name="valid_until" value="${esc(c.valid_until || "")}"></label></div><label class="dossier-checkbox"><input type="checkbox" name="unresolved" ${c.temporal_scope === "unresolved" ? "checked" : ""}> La fecha está pendiente de aclarar</label>${fact?.status === "declared" ? '<label>Tipo de cambio<select name="change_kind"><option value="historical">Corregir lo guardado, también para resultados anteriores</option><option value="future">Cambiar desde la fecha de inicio indicada</option></select></label>' : ""}<p class="field-help">Guardar expresa tu declaración. Puedes corregirla o retirarla después.</p><p id="fact-error" role="alert"></p><div class="dossier-actions"><button class="button primary" type="submit">Guardar información</button><button class="button secondary" type="button" id="fact-cancel">Cancelar</button></div></form></section>`;
    editor.scrollIntoView({block: "start", behavior: "smooth"});
    editor.querySelector("textarea").focus({preventScroll: true});
    document.querySelector("#fact-cancel").onclick = () => editor.replaceChildren();
    const form = document.querySelector("#fact-form");
    let pending = null;
    form.onsubmit = async (event) => {
      event.preventDefault();
      const values = Object.fromEntries(new FormData(form));
      const [scope, scopeId] = values.scope.split(":");
      const content = {...c, kind: values.kind, statement: values.statement.trim(), scope, scope_id: scopeId || null, valid_from: values.valid_from || null, valid_until: values.valid_until || null, temporal_scope: values.unresolved ? "unresolved" : values.valid_from || values.valid_until ? "dated" : "unspecified"};
      if (!fact) content.topic = "owner_" + values.concept.normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase().replace(/[^a-z0-9]+/g, "_").slice(0,70);
      if (content.kind !== "result_reference") content.result_id = null;
      const body = {business_id: business, action: fact ? "correct" : "declare", content, ...(fact ? {fact_id: fact.fact_id, expected_revision: fact.revision} : {}), change_kind: values.change_kind || "historical", original_text: content.statement};
      const signature = JSON.stringify(body);
      if (!pending || pending.signature !== signature) pending = {signature, key: crypto.randomUUID()};
      const button = form.querySelector("button[type=submit]"); button.disabled = true;
      try {
        await api("/api/business/memory", {method: "POST", body: {...body, request_key: pending.key}});
        if (generation !== state.generation) return;
        toast("Información guardada en la memoria compartida."); await refresh("information");
      } catch (error) { if (form.isConnected) form.querySelector("#fact-error").textContent = error.message; button.disabled = false; }
    };
  };
  render(tab);
}
function datasetMarkup(dossier) {
  return `<h2>Datos y archivos</h2><p>Sube datos para futuras preguntas. Los archivos se mantienen separados y los informes se conservan con la versión utilizada.</p><div class="dossier-cards">${dossier.datasets.map(d => `<article class="dossier-card"><div class="dossier-meta"><strong>${esc(d.title)} · v${d.version}</strong><span class="badge ${d.status !== "ready" || d.corrected ? "amber" : "quiet"}">${d.corrected ? "Corregida: no reutilizar" : d.superseded_by ? "Versión anterior" : d.status === "ready" ? "Disponible" : d.status === "importing" ? "Preparación interrumpida" : "No disponible"}</span></div><p class="field-help">Periodo declarado: ${esc(d.period_from || "Sin inicio indicado")} → ${esc(d.period_until || "Sin fin indicado")}</p>${(d.files || []).map(f => `<p>${esc(f.name)}${f.rows != null ? ` · ${f.rows} filas` : ""} <a href="/api/datasets/file/${esc(f.id)}">Descargar original</a></p>`).join("")}${d.superseded_by ? '<p class="field-help">Hay una versión posterior. Los resultados anteriores no se han recalculado.</p>' : ""}${d.status === "ready" && !d.corrected ? `<button class="button secondary" data-ask-dataset="${esc(d.id)}">Preguntar con esta versión</button>` : ""}</article>`).join("") || '<p class="dossier-empty">Todavía no hay datos. Puedes conversar sobre tu negocio y añadir un CSV cuando lo necesites.</p>'}</div><section class="dossier-card"><h2>Añadir un CSV</h2><form id="dataset-form" class="dossier-form"><label>Nombre del conjunto<input name="title" required maxlength="160" placeholder="Ventas de septiembre"></label><label>Archivo CSV UTF-8<input type="file" name="file" accept=".csv,text/csv" required></label><label>Cómo se relaciona con tus datos<select name="mode"><option value="separate">Conjunto independiente</option><option value="update">Nueva versión</option><option value="correction">Corregir una versión</option></select></label><p class="field-help" id="dataset-mode-help"></p><label id="previous-label" hidden>Versión que sustituye<select name="previous_id"><option value="">Selecciona una versión</option>${dossier.datasets.filter(d => d.status === "ready" && !d.superseded_by && !d.corrected).map(d => `<option value="${esc(d.id)}">${esc(d.title)} · v${d.version}</option>`).join("")}</select></label><div class="dossier-dates"><label>Inicio del periodo (opcional)<input type="date" name="period_from"></label><label>Fin del periodo (opcional)<input type="date" name="period_until"></label></div><p class="field-help">Si no conoces la relación o el periodo, deja el archivo separado. No se detecta ni se resuelve automáticamente el solapamiento entre filas.</p><p id="dataset-result" role="status"></p><button class="button primary" type="submit">Guardar datos</button></form></section>`;
}
function bindDatasetUpload(dossier, business, generation, refresh) {
  const form = document.querySelector("#dataset-form"), output = document.querySelector("#dataset-result");
  const storageKey = "dr-dataset-upload-" + business;
  const saved = store.get(storageKey, null);
  if (saved?.metadata?.previous_id && !Array.from(form.elements.previous_id.options).some(o => o.value === saved.metadata.previous_id)) {
    const option = document.createElement("option"); option.value = saved.metadata.previous_id; option.textContent = "Versión del envío anterior (recuperar reintento)"; form.elements.previous_id.add(option);
  }
  if (saved?.metadata) for (const name of ["title", "mode", "previous_id", "period_from", "period_until"]) form.elements[name].value = saved.metadata[name] || "";
  const mode = form.elements.mode;
  const updateMode = () => { document.querySelector("#previous-label").hidden = mode.value === "separate"; form.elements.previous_id.required = mode.value !== "separate"; document.querySelector("#dataset-mode-help").textContent = {separate: "Se guarda por separado. No se suma ni combina con otros archivos.", update: "Sustituye la versión para nuevas consultas. Los resultados anteriores se conservan como históricos y no se recalculan.", correction: "Corrige errores de la versión elegida. Sus resultados afectados dejan de estar disponibles hasta calcular y revisar de nuevo."}[mode.value]; };
  mode.onchange = updateMode; updateMode();
  form.onsubmit = async (event) => {
    event.preventDefault();
    const file = form.elements.file.files[0];
    const values = Object.fromEntries(new FormData(form));
    const metadata = {business_id: business, title: values.title, mode: values.mode, previous_id: values.mode === "separate" ? "" : values.previous_id, period_from: values.period_from, period_until: values.period_until};
    const button = form.querySelector("button[type=submit]"); button.disabled = true; output.textContent = "Preparando el CSV…";
    try {
      const hash = Array.from(new Uint8Array(await crypto.subtle.digest("SHA-256", await file.arrayBuffer()))).map(v => v.toString(16).padStart(2,"0")).join("");
      const signature = JSON.stringify([metadata, file.name, hash]);
      let pending = store.get(storageKey, null);
      if (!pending || pending.signature !== signature) pending = {signature, metadata, key: crypto.randomUUID()};
      store.set(storageKey, pending);
      const body = new FormData(); body.append("metadata", JSON.stringify({...metadata, request_key: pending.key})); body.append("file", file);
      const result = await api("/api/datasets", {method: "POST", body});
      store.remove(storageKey);
      if (generation !== state.generation) return;
      await refresh();
      const status = document.querySelector("#dataset-result"); if (status) status.textContent = result.message;
    } catch (error) { if (output.isConnected) output.textContent = error.message; button.disabled = false; }
  };
  document.querySelectorAll("[data-ask-dataset]").forEach(button => {
    button.onclick = () => {
      const dataset = dossier.datasets.find(d => d.id === button.dataset.askDataset);
      store.set(questionContextKey(), {analysis_id:dataset.id, label:`${dataset.title} · Versión ${dataset.version}`});
      location.hash = "#ask";
    };
  });
}

File: decision_room/web/dossier.py
"""Owner-facing shared memory and independent, versioned CSV imports."""
import hashlib
from datetime import date
from pathlib import Path
from tempfile import TemporaryDirectory

from psycopg.types.json import Jsonb
from pydantic import ValidationError

from ..database import connect
from .. import service as ingestion
from ..memory import service as memory, context
from ..storage import Storage
from .errors import WebError, identifier, bounded
from .business import profile


def guard(ws, data):
    business = ws.business_id()
    if not isinstance(data, dict) or not business or identifier(data.get('business_id')) != business:
        raise WebError('El negocio activo ha cambiado. Recarga la página.', 409)
    return business


def listing(ws):
    business = ws.business_id()
    if not business:
        raise WebError('Selecciona un negocio.', 409)
    with connect(ws.config) as db, db.transaction():
        memory.lock(db, business)
        facts = memory.current(db, business)
        history = db.execute('''SELECT r.*,s.origin_key,s.payload->>'kind' AS origin_kind,
            s.payload->>'text' AS original_text,s.payload->>'question' AS question,
            (SELECT t.conversation_id FROM chat_turns t WHERE t.business_id=r.business_id AND t.memory_source_id=r.source_id LIMIT 1) AS conversation_id
            FROM memory_revisions r JOIN memory_sources s ON s.id=r.source_id
            WHERE r.business_id=%s ORDER BY r.business_revision DESC''', (business,)).fetchall()
        origins = {(r['fact_id'], r['revision']): r for r in history}
        facts = [origins[(r['fact_id'], r['revision'])] for r in facts]
        datasets = db.execute('''SELECT a.id,a.title,a.status,a.created_at,
            coalesce(v.dataset_id,a.id) AS dataset_id,coalesce(v.version,1) AS version,
            v.period_from,v.period_until,v.superseded_by,coalesce(v.corrected,false) AS corrected,
            (SELECT jsonb_agg(jsonb_build_object('id',s.id,'name',s.original_names->>0,
                'status',s.status,'rows',p.row_count)) FROM sources s
                LEFT JOIN prepared_tables p ON p.source_id=s.id WHERE s.analysis_id=a.id) AS files
            FROM analyses a LEFT JOIN dataset_versions v ON v.analysis_id=a.id
            WHERE a.business_id=%s AND NOT EXISTS (SELECT 1 FROM dataset_uploads u WHERE u.business_id=a.business_id AND u.result->>'pending'='true' AND u.result->>'existing_id' IS NULL AND u.result->>'batch_sha256'=a.batch_sha256) ORDER BY a.created_at DESC,a.id''', (business,)).fetchall()
        return dict(business_id=business, business=profile(db, business), memory=memory.status(ws.config, business),
                    facts=facts, history=history, datasets=datasets)


def change(ws, data):
    business = guard(ws, data)
    allowed = {'business_id', 'action', 'request_key', 'content', 'fact_id', 'expected_revision',
               'original_text', 'reason', 'change_kind'}
    if set(data) - allowed or not {'action', 'request_key'} <= set(data):
        raise WebError('Operación de memoria no válida.')
    try:
        return memory.change(ws.config, business, **{k: v for k, v in data.items() if k != 'business_id'})
    except memory.MemoryError as error:
        raise WebError(str(error), error.status) from None
    except ValidationError:
        raise WebError('Revisa la información, el ámbito y las fechas del recuerdo.') from None


def available(db, business, analysis):
    return db.execute('''SELECT 1 FROM analyses a LEFT JOIN dataset_versions v ON v.analysis_id=a.id
        WHERE a.id=%s AND a.business_id=%s AND a.status='ready' AND NOT coalesce(v.corrected,false) AND NOT EXISTS (SELECT 1 FROM dataset_uploads u WHERE u.business_id=a.business_id AND u.result->>'pending'='true' AND u.result->>'existing_id' IS NULL AND u.result->>'batch_sha256'=a.batch_sha256)''',
        (analysis, business)).fetchone() is not None


def upload(ws, data, filename, content):
    business = guard(ws, data)
    key = identifier(data.get('request_key'))
    title = bounded(data.get('title', ''), 'el nombre del conjunto', 160)
    mode = data.get('mode', 'separate')
    if mode not in ('separate', 'update', 'correction'):
        raise WebError('Selecciona datos independientes, actualización o corrección.')
    previous = identifier(data.get('previous_id')) if mode != 'separate' else None
    if mode == 'separate' and data.get('previous_id'):
        raise WebError('Los datos independientes no sustituyen otro conjunto.')
    try:
        start = date.fromisoformat(data['period_from']) if data.get('period_from') else None
        end = date.fromisoformat(data['period_until']) if data.get('period_until') else None
        if start and end and start > end:
            raise ValueError()
    except (ValueError, TypeError):
        raise WebError('Revisa el periodo: utiliza fechas válidas y ordenadas.') from None
    if not isinstance(filename, str) or '/' in filename or '\\' in filename or any(ord(c) < 32 for c in filename):
        raise WebError('El nombre del archivo no es válido.')
    filename = bounded(filename, 'el nombre del archivo', 180)
    if not filename.lower().endswith('.csv') or not content or len(content) > 20 * 1024**2:
        raise WebError('Sube un CSV UTF-8 con datos, de hasta 20 MB.')
    try:
        content.decode('utf-8-sig')
    except UnicodeDecodeError:
        raise WebError('Guarda el CSV con codificación UTF-8.') from None
    signature = memory.digest([title, filename, mode, str(previous), str(start), str(end), hashlib.sha256(content).hexdigest()])
    # Serialize uploads for this business across ingestion's separate transactions.
    # An interrupted import is recovered by the existing content-addressed importer.
    with connect(ws.config) as db:
        db.execute('SELECT pg_advisory_lock(hashtextextended(%s, 25))', (str(business),))
        saved = db.execute('SELECT * FROM dataset_uploads WHERE business_id=%s AND request_key=%s', (business, key)).fetchone()
        if saved:
            if saved['signature'] != signature:
                raise WebError('Este envío ya se guardó con otros datos.', 409)
            if not saved['result'].get('pending'):
                return saved['result']
        unfinished = db.execute("SELECT 1 FROM dataset_uploads WHERE business_id=%s AND request_key<>%s AND result->>'pending'='true'", (business, key)).fetchone()
        if unfinished:
            raise WebError('Hay una carga interrumpida. Vuelve a enviar el archivo anterior con los mismos datos para recuperarla.', 409)
        old = None
        if previous:
            old = db.execute('''SELECT a.id,coalesce(v.dataset_id,a.id) AS dataset_id,coalesce(v.version,1) AS version,
                v.superseded_by FROM analyses a LEFT JOIN dataset_versions v ON v.analysis_id=a.id
                WHERE a.id=%s AND a.business_id=%s AND a.status='ready' ''', (previous, business)).fetchone()
            if not old or old['superseded_by']:
                raise WebError('La versión seleccionada ya ha cambiado o no está disponible. Recarga la ficha.', 409)
        batch_hash, _ = ingestion.batch_metadata([hashlib.sha256(content).hexdigest()])
        existing = db.execute('SELECT id FROM analyses WHERE business_id=%s AND batch_sha256=%s', (business, batch_hash)).fetchone()
        if existing and previous and existing['id'] != previous and (not saved or saved['result'].get('existing_id')):
            raise WebError('Este archivo ya pertenece a otra versión o conjunto. Reutilízalo desde la ficha.', 409)
        if not saved:
            pending = dict(pending=True, existing_id=str(existing['id']) if existing else None, batch_sha256=batch_hash)
            db.execute('INSERT INTO dataset_uploads(business_id,request_key,signature,result) VALUES (%s,%s,%s,%s)',
                       (business, key, signature, Jsonb(pending)))
        else:
            pending = saved['result']
        if existing and pending['existing_id']:
            if previous and existing['id'] != previous:
                raise WebError('Este archivo ya pertenece a otra versión o conjunto. Reutilízalo desde la ficha.', 409)
            result = dict(analysis_id=str(existing['id']), reused=True,
                          status=ingestion.describe(ws.config, business, existing['id'])['analysis']['status'])
        else:
            with TemporaryDirectory(prefix='dr-csv-') as directory:
                path = Path(directory) / filename
                path.write_bytes(content)
                imported = (ingestion.resume(ws.config, business, existing['id']) if existing else
                            ingestion.import_batch(ws.config, business, [path], title=title))
            result = dict(analysis_id=str(imported['analysis']['id']), reused=False, status=imported['analysis']['status'])
        analysis = identifier(result['analysis_id'])
        with db.transaction():
            memory.lock(db, business)
            if result['status'] == 'ready' and not result['reused']:
                if old:
                    db.execute('''INSERT INTO dataset_versions(business_id,analysis_id,dataset_id,version)
                        VALUES (%s,%s,%s,1) ON CONFLICT (analysis_id) DO NOTHING''', (business, previous, old['dataset_id']))
                db.execute('''INSERT INTO dataset_versions(business_id,analysis_id,dataset_id,version,period_from,period_until)
                    VALUES (%s,%s,%s,%s,%s,%s)''',
                    (business, analysis, old['dataset_id'] if old else analysis, old['version'] + 1 if old else 1, start, end))
                if old:
                    db.execute('UPDATE dataset_versions SET superseded_by=%s,corrected=%s WHERE analysis_id=%s',
                               (analysis, mode == 'correction', previous))
                    if mode == 'correction':
                        context.invalidate(db, business)
            result['message'] = ('Este archivo ya estaba guardado; conserva su conjunto, periodo y versión.' if result['reused'] else
                'Datos disponibles. Los informes anteriores no se han recalculado.' if result['status'] == 'ready' else
                'No se ha podido preparar el CSV. Corrige sus encabezados o filas y vuelve a subirlo. Se conservan los datos anteriores.')
            db.execute('UPDATE dataset_uploads SET result=%s WHERE business_id=%s AND request_key=%s',
                       (Jsonb(result), business, key))
        return result


def download(ws, source_id):
    with connect(ws.config) as db:
        source = db.execute('SELECT * FROM sources WHERE id=%s AND business_id=%s',
                            (identifier(source_id), ws.business_id())).fetchone()
    if not source:
        raise WebError('Archivo no encontrado.', 404)
    content = Storage(ws.config.storage).path(ws.business_id(), source['original_key']).read_bytes()
    if hashlib.sha256(content).hexdigest() != source['sha256']:
        raise WebError('El original no supera la comprobación de integridad.', 409)
    return source['original_names'][0], content
